design-proposals/vm-in-a-container.md

```diff
@@ -0,0 +1,186 @@
+start the VM in.
+This should be as transparent as possible to libvirt.
+
+## Implementation
+
+We can create a small binary which will be invoked by libvirt as custom
+emulator. This binay makes sure the VM is started in the right cgroups and
+namespaces of the target container. This binary will have the `suid` bit set.
+
+To find out the one shim process associated to a VM, we can run the shim
+process with the uuid and the vm name in the commandline like this:
+
+```bash
+virt-launcher -kubevirt.vm.uuid 1234-5678-1234-1234 -kubevirt.vm.name testvm
+```
+
+Further virt-handler can for foward the same information to the emulator binary by
+passing additinoal qemu environment variables:
+
+```xml
+<domain type='qemu' xmlns:qemu='http://libvirt.org/schemas/domain/qemu/1.0'>
+<devices>
+    <emulator>/usr/local/bin/myemulator</emulator>
+</devices>
+<qemu:commandline>
+   <qemu:env name='kubevirt.io.vm.uuid' value='1234-5678-1234-1234'/>
+   <qemu:env name='kubevirt.io.vm.name' value='testvm'/>
+</qemu:commandline>
+</domain>
+```
+
+Note that we have to specify the qemu namespace to use these features.
+
+Based on this information, the binary can check all runninig processes for the
+matching shim process and derive namespaces and cgroups from it. Then it can
+start the real qemu process inside them.  For namespaces it can use the
+`setns()` system call.
+
+### Security
+
+Since the driver binary will have the `suid` bit set, to do all necessary
+operations. This imposes one big security risk: The binary can start
+applications in other namespaces and cgroups (even the one of the host).
+
+A few things can be done to reduce this risk:
+
+ 1. Only allow the qemu user to execute the binary
+ 2. Introduce a shared secret between the VM target container and kubevirt
+ 3. This secret can be unique per pod creation and only valid for a specific
+    amount of time
+ 4. The binary checks that  the secret is present in the kubernetes metadata
+    file before it starts something in that container 
+
+The secret can be delivered to the container through 
+[Kubernetes secrets](https://kubernetes.io/docs/user-guide/secrets/). The
+binary can receive the secret via libvirts <qemu:commandline/> tag as
+environment variable:
+
+```xml
+<qemu:commandline>
+   <qemu:env name='kubevirt.io.secret' value='3098tFJoswfwkjp4'/>
+</qemu:commandline>
+```
+
+Note that this means that the secret is exposed in the Domain XML. If we think
+that having a secret lifespan of e.g. one minute is not secure enough, the
+secret could be encrypted.
+
+To further improve security, a [libvirt
+hook](https://libvirt.org/hooks.html#qemu) will check if the Domain XML has an
+emulator tag which points to the right binary.
+
+### Flow
+
+Pseudocode:
+
+```
+# Getting the pid os container runtime independent
+pid = $(ps aux  | grep virt-launcher | grep "-kubevirt.vm.uuid 1234-5678-1234-1234 | tr -s ' ' | cut -d " " -f2)
```

Comment:
Being picky: It would probably not work with rkt stage2-lkvm runtime, because there the launcher would be run in a VM.